### PR TITLE
[Coin3D] Add new ports for simage, Coin, and SoQt

### PR DESCRIPTION
**Pull request by rickertm**
**Describe the pull request**

- What does your PR fix? Fixes issue #4295

This PR adds ports for the Coin3D libraries simage, Coin, and SoQt.

- Which triplets are supported/not supported? Have you updated the CI baseline?

The simage port should work on all triplets.

The Coin and SoQt ports do not support ARM and UWP builds due to OpenGL. The CI baseline has been updated accordingly.

The x64-osx support for SoQt requires a fix for issue #9879.

Coin requires a dependency on boost-math until issue #8753 is addressed.

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?

Yes.

### CODE REVIEW

File: ports/coin/CONTROL [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Source: coin
+Version: 4.0.0
+Description: A high-level 3D visualization library with Open Inventor 2.1 API
```
Comment (JackBoosY) [inline ports/coin/CONTROL]: Can you add homepage here?

File: ports/coin/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,52 @@
+if(VCPKG_TARGET_ARCHITECTURE STREQUAL "arm" OR VCPKG_TARGET_ARCHITECTURE STREQUAL "arm64")
+    message(FATAL_ERROR "Error: ARM builds are not supported.")
+endif()
+
+if(VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsStore" OR VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsPhone")
+    message(FATAL_ERROR "Error: UWP builds are not supported.")
+endif()
```
Comment (JackBoosY) [inline ports/coin/portfile.cmake]: Can you replace these 7 lines with `vcpkg_fail_port_install(ON_TARGET "UWP" ON_ARCH "arm" "arm64")`?

File: ports/coin/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+include(vcpkg_common_functions)
```
Comment (JackBoosY) [inline ports/coin/portfile.cmake]: We don't need to include this, please remove this line.

File: ports/coin/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        -DCOIN_BUILD_SHARED_LIBS=${COIN_BUILD_SHARED_LIBS}
+        -DCOIN_BUILD_TESTS=OFF
+)
+
+vcpkg_install_cmake()
+vcpkg_copy_pdbs()
+vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/Coin-4.0.0)
+
+file(COPY ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/coin)
+file(RENAME ${CURRENT_PACKAGES_DIR}/share/coin/COPYING ${CURRENT_PACKAGES_DIR}/share/coin/copyright)
```
Comment (JackBoosY) [inline ports/coin/portfile.cmake]: We prefer to use `file(INSTALL ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)` here.

File: ports/simage/CONTROL [hunk 5]
```diff
@@ -0,0 +1,3 @@
+Source: simage
+Version: 1.8.0
+Description: Image file format library abstraction layer
```
Comment (JackBoosY) [inline ports/simage/CONTROL]: Can you add homepage here?

File: ports/simage/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,47 @@
+include(vcpkg_common_functions)
```
Comment (JackBoosY) [inline ports/simage/portfile.cmake]: We don't need to include this, please remove this line.

File: ports/simage/portfile.cmake [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        -DSIMAGE_USE_AVIENC=${SIMAGE_USE_AVIENC}
+        -DSIMAGE_USE_GDIPLUS=${SIMAGE_USE_GDIPLUS}
+)
+
+vcpkg_install_cmake()
+vcpkg_copy_pdbs()
+vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/simage-1.8.0)
+
+file(COPY ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/simage)
+file(RENAME ${CURRENT_PACKAGES_DIR}/share/simage/LICENSE ${CURRENT_PACKAGES_DIR}/share/simage/copyright)
```
Comment (JackBoosY) [inline ports/simage/portfile.cmake]: We prefer to use `file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)` here.

File: ports/soqt/CONTROL [hunk 8]
```diff
@@ -0,0 +1,4 @@
+Source: soqt
+Version: 1.6.0
+Description: A Qt Gui-toolkit binding for Coin
```
Comment (JackBoosY) [inline ports/soqt/CONTROL]: Can you add homepage here?

File: ports/soqt/portfile.cmake [hunk 9]
```diff
@@ -0,0 +1,98 @@
+if(VCPKG_TARGET_ARCHITECTURE STREQUAL "arm" OR VCPKG_TARGET_ARCHITECTURE STREQUAL "arm64")
+    message(FATAL_ERROR "Error: ARM builds are not supported.")
+endif()
+
+if(VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsStore" OR VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsPhone")
+    message(FATAL_ERROR "Error: UWP builds are not supported.")
+endif()
```
Comment (JackBoosY) [inline ports/soqt/portfile.cmake]: Can you replace these 7 lines with `vcpkg_fail_port_install(ON_TARGET "UWP" ON_ARCH "arm" "arm64")`?

File: ports/soqt/portfile.cmake [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+include(vcpkg_common_functions)
```
Comment (JackBoosY) [inline ports/soqt/portfile.cmake]: We don't need to include this, please remove this line.

File: ports/soqt/portfile.cmake [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        -DSOQT_BUILD_DOCUMENTATION=OFF
+        -DSOQT_BUILD_SHARED_LIBS=${SOQT_BUILD_SHARED_LIBS}
+)
+
+vcpkg_install_cmake()
+vcpkg_copy_pdbs()
+vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/SoQt-1.6.0)
+
+file(COPY ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/soqt)
+file(RENAME ${CURRENT_PACKAGES_DIR}/share/soqt/COPYING ${CURRENT_PACKAGES_DIR}/share/soqt/copyright)
```
Comment (JackBoosY) [inline ports/soqt/portfile.cmake]: We prefer to use `file(INSTALL ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)` here.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/simage/portfile.cmake [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        potentially-uninitialized-local-pointer-variable.patch
+)
+
+if(VCPKG_LIBRARY_LINKAGE STREQUAL static)
+    set(SIMAGE_BUILD_SHARED_LIBS OFF)
+else()
+    set(SIMAGE_BUILD_SHARED_LIBS ON)
+endif()
+
+if(VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsStore" OR VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsPhone")
```
Comment (NancyLi1013) [inline ports/simage/portfile.cmake]: Could you please update this line as `if(VCPKG_TARGET_IS_UWP)`?

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by rickertm** [comment]
Thank you all for the feedback. Should be integrated now.

**Comment by JackBoosY** [comment]
I found some linux regressions, please check the logs [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=27772&view=artifacts&type=publishedArtifacts).

**Comment by rickertm** [comment]
The error log states, that the file `/ci/myagent/_work/1/s/installed/x64-linux/share/cmake/Qt5Gui/Qt5GuiConfigExtras.cmake` of package qt5-base contains a reference to file `/ci/myagent/_work/3/s/installed/x64-linux/lib/libEGL.so` that apparently does not exist on disk.

Apart from the reference to the missing `libEGL.so` file in `/ci/myagent/_work/3/...`, all other entries in the log file point to the directory `/ci/myagent/_work/1/...`?

I've just tested this again on a local Ubuntu 18.04 installation and it compiled without issues.

**Comment by JackBoosY** [comment]
@rickertm I will take a look.

**Comment by JackBoosY** [comment]
I can't repro that regression too, so I rerun the pipeline test.

**Comment by JackBoosY** [comment]
@ras0219-msft Could you have a look?

**Comment by rickertm** [comment]
The issue with the missing file `/ci/myagent/_work/3/s/installed/x64-linux/lib/libEGL.so` seems to have been addressed in #10390?

**Comment by JackBoosY** [comment]
@rickertm We are trying to resolve that.

**Comment by JackBoosY** [comment]
Please resolve linx regression, get log [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=27909&view=artifacts&type=publishedArtifacts).

**Comment by rickertm** [comment]
Same issue as before, file `libEGL.so` is still missing and references folder `/ci/myagent/_work/2/...` instead of `/ci/myagent/_work/3/...`
```
CMake Error at /ci/myagent/_work/3/s/installed/x64-linux/share/cmake/Qt5Gui/Qt5GuiConfig.cmake:15 (message):
  The imported target "Qt5::Gui" references the file

     "/ci/myagent/_work/2/s/installed/x64-linux/lib/libEGL.so"

  but this file does not exist.  Possible reasons include:
```

**Comment by rickertm** [comment]
Seems to work now after rebase to master.

**Review by JackBoosY — APPROVED** [review]

File: ports/coin/CONTROL [hunk 13]
```diff
@@ -0,0 +1,5 @@
...
+Build-Depends: boost-assert, boost-config, boost-lexical-cast, boost-math, boost-smart-ptr, boost-static-assert, simage
+Homepage: https://github.com/coin3d/coin
```
Comment (NancyLi1013) [inline ports/coin/CONTROL]: Could you please add `Supports: !(arm|arm64|uwp)`?

File: ports/soqt/CONTROL [hunk 14]
```diff
@@ -0,0 +1,5 @@
...
+Build-Depends: coin, qt5-base
+Homepage: https://github.com/coin3d/soqt
```
Comment (NancyLi1013) [inline ports/soqt/CONTROL]: Could you pleae add `Supports: !(arm|arm64|uwp)`?

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by rickertm** [comment]
Done, thank you for the suggestion.

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
LGTM. Thanks for your PR.

**Comment by strega-nil** [comment]
@rickertm what's up with the downloading distfiles for `soanydata` and `sogui`, as opposed to setting up ports for them by themselves? I don't think I understand.

**Comment by rickertm** [comment]
`soanydata` and `sogui` are submodules that were not included in the tarball archive at the time of this PR, so I followed the canonical way for this case described in https://github.com/microsoft/vcpkg/issues/1036.

The upstream project has since added additional tarball archives that include the submodules, I can update the PR accordingly if you prefer. The patches have also been integrated upstream in the meantime, but there is no new tagged release version yet.

**Comment by strega-nil** [comment]
Ah, cool, that's fine then :) just curious.

**Comment by rickertm** [comment]
I've added additional patches for simage when linking against libsndfile (FLAC and Ogg/Vorbis order). I've also disabled examples and tests. All patches have been submitted upstream.

On Linux and UPW, the simage port now depends on giflib, libpng etc., as VfW, GDI+, or CGImage are not available.

The Coin port adds features for optional libraries that are linked at runtime, with simage and zlib as default features.

**Comment by rickertm** [comment]
The boost-math dependency in Coin can be removed once #11131 is merged.

**Comment by JackBoosY** [comment]
`libqcow:x64-linux` regression are not related with this PR.

**Comment by strega-nil** [comment]
Thanks so much @rickertm :D

**Comment by rickertm** [comment]
Thank you for merging this @strega-nil.

**Pull request merged**